opm/common/utility/CSRGraphFromCoordinates.hpp

```diff
@@ -0,0 +1,491 @@
+#include <vector>
+
+/// \file
+///
+/// Facility for converting collection of region ID pairs into a sparse
+/// (CSR) adjacency matrix representation of a graph.  Supports O(nnz)
+/// compression.
+
+namespace Opm { namespace utility {
+
+    /// Form CSR adjacency matrix representation of unstructured graphs.
+    /// Optionally maps vertex pairs to compressed indices to enable O(1)
+    /// per-element lookup in assembly-like operations.
+    ///
+    /// \tparam VertexID ID type of an abstract vertex in the graph.  Could
+    ///    for instance be used to represent a cell index or a region index.
+    ///    Must be an integral type.
+    ///
+    /// \tparam TrackCompressedIdx Whether or not to form a mapping relation
+    ///    for vertex pairs to compressed indices.  Default value, false,
+    ///    bypasses this mapping relation and conserves memory.
+    template <typename VertexID = int, bool TrackCompressedIdx = false>
+    class CSRGraphFromCoordinates
+    {
+    private:
+        using BaseVertexID = std::remove_cv_t<VertexID>;
+
+        static_assert(std::is_integral_v<BaseVertexID>,
+                      "The VertexID must be an integral type");
+
+        /// Representation of neighbouring regions.
+        using Neighbours = std::vector<BaseVertexID>;
+
+        /// Offset into neighbour array.
+        using Offset = typename Neighbours::size_type;
+
+        /// CSR start pointers.
+        using Start = std::vector<Offset>;
+
+    public:
+        /// Clear all internal buffers, but preserve allocated capacity.
+        void clear();
+
+        /// Add flow rate connection between regions.
+        ///
+        /// \param[in] v1 First vertex in vertex pair.  Used as row index.
+        ///
+        /// \param[in] r2 Second vertex in vertex pair.  Used as column index.
+        ///
+        /// If both vertex IDs are the same then this function does nothing.
+        void addConnection(VertexID v1, VertexID v2);
+
+        /// Form CSR adjacency matrix representation of input graph from
+        /// connections established in previous calls to addConnection().
+        ///
+        /// \param[in] maxNumVertices Number of rows in resulting CSR
+        ///     matrix.  If prior calls to addConnection() supply vertex IDs
+        ///     (row indices) greater than or equal to \p maxNumVertices,
+        ///     then method compress() will throw \code
+        ///     std::invalid_argument \endcode.
+        void compress(Offset maxNumVertices);
+
+        /// Retrieve number of rows (source entities) in input graph.
+        /// Corresponds to value of argument passed to compress().  Valid
+        /// only after calling compress().
+        Offset numVertices() const;
+
+        /// Retrieve number of edges (non-zero matrix elements) in input
+        /// graph.
+        Offset numEdges() const;
+
+        /// Read-only access to compressed structure's start pointers.
+        decltype(auto) startPointers() const
```

Comment:
Should we not explicitly return a const reference here? No need to decltype(auto) unless we need to sometimes return a reference, sometimes not.

Reply:
I think a user of the class needs to know if it is a reference or copy that is made, which is hidden by decltype(auto).

Reply:
> I think a user of the class needs to know if it is a reference or copy that is made, which is hidden by `decltype(auto)`.

The only case I can think of that would need to know that is if the user considers calling the function in an inner loop.  Most of the time I prefer to write
```C++
const auto& x = obj.memfun(args);
```
and let type deduction and/or lifetime extension do the rest.  For what it's worth, I still think `decltype(auto)` expresses the intent better here, but I've pushed an update to use the specific reference types instead.